pandas/tests/reshape/test_concat.py

```diff
@@ -829,12 +831,126 @@ def test_append_preserve_index_name(self):
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+    ], r=2), ids=lambda x: str(x.dtype))
+    def test_append_different_columns_types(self, df_columns, series_index):
+        # GH18359
+        # see also tests 'test_append_multi_index_raises' and
+        # 'test_append_interval_index_raises' below
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=df_columns)
+        ser = pd.Series([7, 8, 9], index=series_index, name=2)
+
+        result = df.append(ser)
+        idx_diff = ser.index.difference(df_columns)
+        combined_columns = Index(df_columns.tolist()).append(idx_diff)
+        expected = pd.DataFrame([[1., 2., 3., np.nan, np.nan, np.nan],
+                                 [4, 5, 6, np.nan, np.nan, np.nan],
+                                 [np.nan, np.nan, np.nan, 7, 8, 9]],
+                                index=[0, 1, 2],
+                                columns=combined_columns)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("other_type", [
+        pd.RangeIndex(3),
+        pd.CategoricalIndex('A B C'.split()),
+        pd.CategoricalIndex('A B C'.split(), ordered=True),
+        pd.IntervalIndex.from_breaks([0, 1, 2, 3]),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+        pd.Index([4, 5, 6]),
+    ], ids=lambda x: str(x.dtype))
+    def test_append_multi_index_raises(self, other_type):
+        # GH18359
+        # .append will raise if MultiIndex appends or is appended to a
+        # different index type
+
+        mi = pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()])
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=mi)
+        ser = pd.Series([7, 8, 9], index=other_type, name=2)
+        with pytest.raises(TypeError):
+            df.append(ser)
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=other_type)
+        ser = pd.Series([7, 8, 9], index=mi, name=2)
+        with pytest.raises(TypeError):
+            df.append(ser)
+
+    @pytest.mark.parametrize("other_type", [
+        pd.RangeIndex(3),
+        pd.CategoricalIndex('A B C'.split()),
+        pd.CategoricalIndex('A B C'.split(), ordered=True),
+        pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()]),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+        pd.Index([4, 5, 6]),
+    ], ids=lambda x: str(x.dtype))
+    def test_append_interval_index_raises(self, other_type):
+        # GH18359
+        # .append will raise if IntervalIndex appends or is appended to a
+        # different index type
+
+        ii = pd.IntervalIndex.from_breaks([0, 1, 2, 3])
+
```

Comment:
this tests now looks superfluous, yes?  and you just need to add it to the above

Reply:
This still fails, only with a TypeError rather than a ValueError, so not sure I understand.

Reply:
yes. None of these should fail, you are catching the ``TypeError`` already (and then coercing to ``object``).

Reply:
Well, combining IntervalIndex and other index types will stilll fail, because ``Interval`` and other dtypes are not orderable:

```python
>>> ii = pd.IntervalIndex.from_breaks([0,1,2,3])
>>> i = pd.RangeIndex(3)
>>> df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=ii)
>>> ser = pd.Series([7, 8, 9], index=i, name=2)
>>> df.append(ser)
TypeError: unorderable types: Interval() > int()
```

``test_append_multi_index_raises`` and ``test_append_interval_index_raises`` are very similar and I've made a changes to combine them into a single test, see ``test_append_different_columns_types_raises``.